process/datadog_checks/process/process.py

```diff
@@ -109,63 +213,72 @@ def find_pids(self, name, search_string, exact_match, ignore_ad=True):
 
         refresh_ad_cache = self.should_refresh_ad_cache(name)
 
-        encountered_process_names = set()
         matching_pids = set()
 
-        for proc in psutil.process_iter():
-            # Skip access denied processes
-            if not refresh_ad_cache and proc.pid in self.ad_cache:
-                continue
+        # Acquire the write lock to check whether to refresh because we're
+        # going to keep it to do the refresh, AND, we don't want multiple
+        # threads getting a `yes` result at once
+        with self.process_list_cache.write_lock():
+            if self.process_list_cache.should_refresh_proclist():
+                self.log.debug("Refreshing process list")
+                self.process_list_cache.elements = [proc for proc in psutil.process_iter(attrs=['pid', 'name'])]
+                self.process_list_cache.last_ts = time.time()
+                self.log.debug("Set last ts to %s", self.process_list_cache.last_ts)
+            else:
+                self.log.debug("Using process list cache")
```

Comment:
I think we should move this section to a method on `ProcessListCache`.

Maybe move all the new objects to a `cache` module too, what do you think?